File: inspect_audit.js
const fs = require('fs');
const path = require('path');

const root = path.resolve('e:/pi-kangpom-main');
const entry = path.join(root, 'app', 'page.tsx');
const allowedExts = new Set(['.ts', '.tsx', '.js', '.jsx', '.mjs', '.cjs']);

function relpath(p) {
  try { return path.relative(root, p).split(path.sep).join('/'); } catch { return p; }
}

function walk(dir) {
  let files = [];
  for (const entryName of fs.readdirSync(dir, { withFileTypes: true })) {
    const full = path.join(dir, entryName.name);
    if (entryName.isDirectory()) {
      files = files.concat(walk(full));
    } else if (allowedExts.has(path.extname(entryName.name))) {
      files.push(full);
    }
  }
  return files;
}

const allCodeFiles = walk(root).map(f => path.resolve(f));
const fileSet = new Set(allCodeFiles);

const tsconfigPath = path.join(root, 'tsconfig.json');
const paths = {};
if (fs.existsSync(tsconfigPath)) {
  const tsconfig = JSON.parse(fs.readFileSync(tsconfigPath, 'utf8'));
  for (const [alias, targets] of Object.entries(tsconfig.compilerOptions?.paths || {})) {
    if (alias.endsWith('/*')) {
      const base = alias.slice(0, -1);
      paths[base] = (targets || []).map(t => path.resolve(root, t.endsWith('/*') ? t.slice(0, -1) : t));
    } else {
      paths[alias] = (targets || []).map(t => path.resolve(root, t));
    }
  }
}

function resolveImport(fromFile, spec) {
  if (spec.startsWith('http') || spec.startsWith('mailto:')) return null;
  if (spec.startsWith('.') || spec.startsWith('/')) {
    const candidates = [];
    if (spec.startsWith('/')) {
      const base = path.resolve(root, spec.slice(1));
      candidates.push(base);
    } else {
      const base = path.resolve(path.dirname(fromFile), spec);
      candidates.push(base);
      for (const ext of ['.ts', '.tsx', '.js', '.jsx', '.mjs', '.cjs']) {
        candidates.push(base + ext);
      }
      candidates.push(path.join(base, 'index.ts'));
      candidates.push(path.join(base, 'index.tsx'));
      candidates.push(path.join(base, 'index.js'));
      candidates.push(path.join(base, 'index.jsx'));
      candidates.push(path.join(base, 'index.mjs'));
      candidates.push(path.join(base, 'index.cjs'));
    }
    for (const cand of candidates) {
      if (fs.existsSync(cand) && fs.statSync(cand).isFile() && allowedExts.has(path.extname(cand))) return path.resolve(cand);
      if (fs.existsSync(cand) && fs.statSync(cand).isDirectory()) {
        for (const idx of [path.join(cand, 'index.ts'), path.join(cand, 'index.tsx'), path.join(cand, 'index.js'), path.join(cand, 'index.jsx'), path.join(cand, 'index.mjs'), path.join(cand, 'index.cjs')]) {
          if (fs.existsSync(idx) && fs.statSync(idx).isFile()) return path.resolve(idx);
        }
      }
    }
    return null;
  }
  for (const [alias, targets] of Object.entries(paths)) {
    if (spec.startsWith(alias)) {
      const suffix = spec.slice(alias.length).replace(/^\//, '');
      for (const target of targets) {
        const base = fs.existsSync(target) && fs.statSync(target).isDirectory() ? path.join(target, suffix) : target;
        if (fs.existsSync(base) && fs.statSync(base).isFile() && allowedExts.has(path.extname(base))) return path.resolve(base);
        for (const ext of ['.ts', '.tsx', '.js', '.jsx', '.mjs', '.cjs']) {
          const test = base + ext;
          if (fs.existsSync(test) && fs.statSync(test).isFile()) return path.resolve(test);
        }
        for (const idx of [path.join(base, 'index.ts'), path.join(base, 'index.tsx'), path.join(base, 'index.js'), path.join(base, 'index.jsx'), path.join(base, 'index.mjs'), path.join(base, 'index.cjs')]) {
          if (fs.existsSync(idx) && fs.statSync(idx).isFile()) return path.resolve(idx);
        }
      }
    }
  }
  return null;
}

const adj = new Map();
const importsMeta = new Map();
const importedBy = new Map();
const reachable = new Set();
const visited = new Set();
const queue = [path.resolve(entry)];

while (queue.length) {
  const f = queue.pop();
  if (visited.has(f)) continue;
  visited.add(f);
  if (!fs.existsSync(f) || !allowedExts.has(path.extname(f))) continue;
  reachable.add(f);
  const text = fs.readFileSync(f, 'utf8');
  const matches = [];
  for (const line of text.split(/\r?\n/)) {
    const stripped = line.trim();
    if (!stripped.startsWith('import ') && !stripped.startsWith('export ')) continue;
    const m = stripped.match(/['"]([^'"]+)['"]/);
    if (!m) continue;
    const spec = m[1];
    if (spec.endsWith('.css')) continue;
    const resolved = resolveImport(f, spec);
    if (resolved === null) {
      matches.push([spec, null, 'unknown', stripped]);
    } else {
      const isType = stripped.includes('import type') || stripped.includes('import { type') || stripped.includes('import type {');
      matches.push([spec, resolved, isType ? 'type' : 'runtime', stripped]);
    }
  }
  importsMeta.set(f, matches);
  for (const [spec, resolved] of matches) {
    if (!resolved || !fileSet.has(resolved)) continue;
    if (!adj.has(f)) adj.set(f, []);
    adj.get(f).push(resolved);
    if (!importedBy.has(resolved)) importedBy.set(resolved, []);
    importedBy.get(resolved).push(f);
    if (!visited.has(resolved)) queue.push(resolved);
  }
}

const cycles = [];
const stack = [];
const stackSet = new Set();
const seen = new Set();
function dfs(node) {
  stack.push(node);
  stackSet.add(node);
  for (const nxt of adj.get(node) || []) {
    if (!reachable.has(nxt)) continue;
    if (stackSet.has(nxt)) {
      const idx = stack.indexOf(nxt);
      cycles.push(stack.slice(idx).concat([nxt]));
    } else if (!seen.has(nxt)) {
      dfs(nxt);
    }
  }
  stack.pop();
  stackSet.delete(node);
  seen.add(node);
}
for (const node of Array.from(reachable).sort()) {
  if (!seen.has(node)) dfs(node);
}
const normalizedCycles = [];
const seenCycles = new Set();
for (const cyc of cycles) {
  const norm = cyc.length <= 1 ? tuple(cyc) : minRotation(cyc);
  if (!seenCycles.has(norm)) {
    seenCycles.add(norm);
    normalizedCycles.push(norm.map(x => x));
  }
}
function tuple(arr) { return JSON.stringify(arr); }
function minRotation(cyc) {
  const reps = [];
  for (let i = 0; i < cyc.length - 1; i++) {
    reps.push(JSON.stringify(cyc.slice(i).concat(cyc.slice(0, i))));
  }
  return JSON.parse(reps.sort()[0]);
}

const interfaceDefs = new Map();
const typeDefs = new Map();
const enumDefs = new Map();
const constantDefs = new Map();
const hookDefs = new Map();
const systemDefs = new Map();
for (const f of reachable) {
  const text = fs.readFileSync(f, 'utf8');
  for (const m of text.matchAll(/\binterface\s+([A-Za-z0-9_]+)\b/g)) {
    if (!interfaceDefs.has(m[1])) interfaceDefs.set(m[1], []); interfaceDefs.get(m[1]).push(f);
  }
  for (const m of text.matchAll(/\btype\s+([A-Za-z0-9_]+)\b/g)) {
    if (!typeDefs.has(m[1])) typeDefs.set(m[1], []); typeDefs.get(m[1]).push(f);
  }
  for (const m of text.matchAll(/\benum\s+([A-Za-z0-9_]+)\b/g)) {
    if (!enumDefs.has(m[1])) enumDefs.set(m[1], []); enumDefs.get(m[1]).push(f);
  }
  for (const m of text.matchAll(/\b(?:const|let|var)\s+([A-Z0-9_]+)\b/g)) {
    if (!constantDefs.has(m[1])) constantDefs.set(m[1], []); constantDefs.get(m[1]).push(f);
  }
  for (const m of text.matchAll(/\b(?:function|const)\s+(use[A-Za-z0-9_]+)\b/g)) {
    if (!hookDefs.has(m[1])) hookDefs.set(m[1], []); hookDefs.get(m[1]).push(f);
  }
  for (const m of text.matchAll(/\b(?:class|function|const|let|var)\s+([A-Za-z0-9_]+(?:System|Game)[A-Za-z0-9_]*)\b/g)) {
    if (!systemDefs.has(m[1])) systemDefs.set(m[1], []); systemDefs.get(m[1]).push(f);
  }
}

const exportsByFile = new Map();
for (const f of reachable) {
  const text = fs.readFileSync(f, 'utf8');
  const exports = [];
  for (const m of text.matchAll(/\bexport\s+(?:default\s+)?(?:async\s+)?(?:function|class|const|let|var|interface|type|enum)\s+([A-Za-z0-9_]+)/g)) exports.push(['decl', m[1]]);
  for (const m of text.matchAll(/\bexport\s*\{([^}]+)\}/g)) {
    for (const part of m[1].split(',')) {
      const trimmed = part.trim();
      if (!trimmed) continue;
      exports.push(['named', trimmed.includes(' as ') ? trimmed.split(' as ')[1].trim() : trimmed]);
    }
  }
  if (/\bexport\s+default\b/.test(text)) exports.push(['default', 'default']);
  exportsByFile.set(f, exports);
}

const unusedExports = [];
for (const [f, exports] of exportsByFile) {
  for (const [kind, name] of exports) {
    let imported = false;
    for (const src of reachable) {
      if (src === f) continue;
      for (const [spec, resolved, depKind, line] of importsMeta.get(src) || []) {
        if (resolved !== f) continue;
        const clause = line.split('from', 1)[0].replace('import ', '', 1).replace('type ', '', 1);
        if (clause.includes('{')) {
          const inner = clause.split('{')[1].split('}')[0];
          for (const part of inner.split(',')) {
            const trimmed = part.trim();
            if (!trimmed) continue;
            const importedName = trimmed.includes(' as ') ? trimmed.split(' as ')[1].trim() : trimmed;
            if (importedName === name || (name === 'default' && importedName === 'default')) { imported = true; break; }
          }
        } else if (clause.trim() && clause.trim() !== '*') {
          if (name === 'default') imported = true;
        }
        if (imported) break;
      }
      if (imported) break;
    }
    if (!imported) unusedExports.push([f, kind, name]);
  }
}

const allImportedTargets = new Set();
for (const src of reachable) {
  for (const [spec, resolved] of importsMeta.get(src) || []) {
    if (resolved && fileSet.has(resolved)) allImportedTargets.add(resolved);
  }
}
const deadFiles = allCodeFiles.filter(f => f !== path.resolve(entry) && !allImportedTargets.has(f));

const allDirs = new Map();
for (const f of allCodeFiles) {
  const dir = path.dirname(f);
  if (!allDirs.has(dir)) allDirs.set(dir, []);
  allDirs.get(dir).push(f);
}
const deadFolders = [];
for (const [d, files] of allDirs) {
  if (files.some(f => reachable.has(f))) continue;
  const relParts = path.relative(root, d).split(path.sep).filter(Boolean);
  if (relParts[0] && ['.next', 'node_modules'].includes(relParts[0])) continue;
  deadFolders.push([d, files.map(f => relpath(f))]);
}

const brokenImports = [];
for (const f of Array.from(reachable).sort()) {
  for (const [spec, resolved, kind, line] of importsMeta.get(f) || []) {
    if (resolved === null) brokenImports.push([f, spec, line]);
  }
}

const out = [];
out.push('RUNTIME_DEPENDENCY_AUDIT');
out.push('ENTRY=' + path.resolve(entry));
out.push('');
out.push('1. COMPLETE DEPENDENCY TREE');
for (const f of Array.from(reachable).sort((a,b)=>relpath(a).localeCompare(relpath(b)))) {
  out.push('FILE=' + relpath(f));
  out.push('ABSOLUTE_PATH=' + f);
  out.push('IMPORTED_BY=' + (importedBy.get(f) || []).map(x => relpath(x)).join(', ') || 'NONE');
  const imports = [];
  for (const [spec, resolved, kind, line] of importsMeta.get(f) || []) {
    if (resolved === null) imports.push(spec + ' [UNRESOLVED]');
    else if (reachable.has(resolved)) imports.push(relpath(resolved));
    else imports.push(spec);
  }
  out.push('IMPORTS=' + (imports.join(', ') || 'NONE'));
  out.push('RUNTIME_USED=YES');
  out.push('');
}
out.push('2. FILES THAT EXIST BUT ARE NEVER IMPORTED ANYWHERE');
for (const f of deadFiles.sort((a,b)=>relpath(a).localeCompare(relpath(b)))) out.push(relpath(f));
out.push('');
out.push('3. EXPORTED SYMBOLS THAT ARE NEVER IMPORTED');
for (const [f, kind, name] of unusedExports.sort((a,b)=>relpath(a[0]).localeCompare(relpath(b[0])) || a[1].localeCompare(b[1]) || a[2].localeCompare(b[2]))) out.push(relpath(f) + ' :: ' + kind + ' :: ' + name);
out.push('');
out.push('4. DUPLICATE INTERFACES');
for (const [name, files] of Array.from(interfaceDefs.entries()).sort((a,b)=>a[0].localeCompare(b[0]))) {
  if (new Set(files).size > 1) { out.push(name); for (const f of Array.from(new Set(files)).sort((a,b)=>relpath(a).localeCompare(relpath(b)))) out.push('  ' + relpath(f)); }
}
out.push('');
out.push('5. DUPLICATE TYPES');
for (const [name, files] of Array.from(typeDefs.entries()).sort((a,b)=>a[0].localeCompare(b[0]))) {
  if (new Set(files).size > 1) { out.push(name); for (const f of Array.from(new Set(files)).sort((a,b)=>relpath(a).localeCompare(relpath(b)))) out.push('  ' + relpath(f)); }
}
out.push('');
out.push('6. DUPLICATE CONSTANTS');
for (const [name, files] of Array.from(constantDefs.entries()).sort((a,b)=>a[0].localeCompare(b[0]))) {
  if (new Set(files).size > 1) { out.push(name); for (const f of Array.from(new Set(files)).sort((a,b)=>relpath(a).localeCompare(relpath(b)))) out.push('  ' + relpath(f)); }
}
out.push('');
out.push('7. DUPLICATE HOOKS');
for (const [name, files] of Array.from(hookDefs.entries()).sort((a,b)=>a[0].localeCompare(b[0]))) {
  if (new Set(files).size > 1) { out.push(name); for (const f of Array.from(new Set(files)).sort((a,b)=>relpath(a).localeCompare(relpath(b)))) out.push('  ' + relpath(f)); }
}
out.push('');
out.push('8. DUPLICATE GAME SYSTEMS');
for (const [name, files] of Array.from(systemDefs.entries()).sort((a,b)=>a[0].localeCompare(b[0]))) {
  if (new Set(files).size > 1) { out.push(name); for (const f of Array.from(new Set(files)).sort((a,b)=>relpath(a).localeCompare(relpath(b)))) out.push('  ' + relpath(f)); }
}
out.push('');
out.push('9. CIRCULAR DEPENDENCY CHAINS');
for (const cyc of normalizedCycles) out.push(cyc.join(' -> '));
out.push('');
out.push('10. FINAL RUNTIME GRAPH');
for (const f of Array.from(reachable).sort((a,b)=>relpath(a).localeCompare(relpath(b)))) {
  const targets = [];
  for (const [spec, resolved, kind] of importsMeta.get(f) || []) {
    if (resolved && reachable.has(resolved) && kind === 'runtime') targets.push(relpath(resolved));
  }
  out.push(relpath(f) + ' -> ' + (targets.join(', ') || 'NONE'));
}
out.push('');
out.push('11. BROKEN IMPORTS / UNRESOLVED PATHS');
for (const [f, spec] of brokenImports) out.push(relpath(f) + ' -> ' + spec + ' [UNRESOLVED]');
out.push('');
out.push('12. DEAD FOLDERS');
for (const [d, files] of deadFolders.sort((a,b)=>relpath(a[0]).localeCompare(relpath(b[0])))) out.push(relpath(d) + ' :: ' + files.join(', '));
console.log(out.join('\n'));

File: inspect_audit.py
import re, json, sys
from pathlib import Path
from collections import defaultdict

root = Path('e:/pi-kangpom-main').resolve()
entry = root / 'app' / 'page.tsx'
allowed_exts = {'.ts', '.tsx', '.js', '.jsx', '.mjs', '.cjs'}

paths = {}
tsconfig = root / 'tsconfig.json'
if tsconfig.exists():
    data = json.loads(tsconfig.read_text(encoding='utf-8'))
    for alias, targets in data.get('compilerOptions', {}).get('paths', {}).items():
        if alias.endswith('/*'):
            base = alias[:-1]
            paths[base] = [str((root / t[:-2]).resolve()) if t.endswith('/*') else str((root / t).resolve()) for t in targets]
        else:
            paths[alias] = [str((root / t).resolve()) for t in targets]

all_code_files = [p.resolve() for p in root.rglob('*') if p.is_file() and p.suffix.lower() in allowed_exts]
file_set = set(all_code_files)


def resolve_import(from_file, spec):
    if spec.startswith('http') or spec.startswith('mailto:'):
        return None
    if spec.startswith('.') or spec.startswith('/'):
        candidates = []
        if spec.startswith('/'):
            base = (root / spec[1:]).resolve()
            candidates.append(base)
        else:
            base = (from_file.parent / spec).resolve()
            candidates.append(base)
            for ext in ['.ts', '.tsx', '.js', '.jsx', '.mjs', '.cjs']:
                candidates.append(base.with_suffix(ext))
            candidates.append(base / 'index.ts')
            candidates.append(base / 'index.tsx')
            candidates.append(base / 'index.js')
            candidates.append(base / 'index.jsx')
            candidates.append(base / 'index.mjs')
            candidates.append(base / 'index.cjs')
        for cand in candidates:
            if cand.exists() and cand.is_file() and cand.suffix.lower() in allowed_exts:
                return cand.resolve()
            if cand.exists() and cand.is_dir():
                for idx in [cand / 'index.ts', cand / 'index.tsx', cand / 'index.js', cand / 'index.jsx', cand / 'index.mjs', cand / 'index.cjs']:
                    if idx.exists() and idx.is_file():
                        return idx.resolve()
        return None
    for alias, targets in paths.items():
        if spec.startswith(alias):
            suffix = spec[len(alias):]
            if suffix.startswith('/'):
                suffix = suffix[1:]
            for target in targets:
                base = Path(target)
                if base.exists() and base.is_dir():
                    cand = base / suffix
                else:
                    cand = base
                if cand.exists() and cand.is_file() and cand.suffix.lower() in allowed_exts:
                    return cand.resolve()
                for ext in ['.ts', '.tsx', '.js', '.jsx', '.mjs', '.cjs']:
                    test = cand.with_suffix(ext)
                    if test.exists() and test.is_file():
                        return test.resolve()
                for idx in [cand / 'index.ts', cand / 'index.tsx', cand / 'index.js', cand / 'index.jsx', cand / 'index.mjs', cand / 'index.cjs']:
                    if idx.exists() and idx.is_file():
                        return idx.resolve()
    return None


def relpath(path):
    try:
        return path.relative_to(root).as_posix()
    except Exception:
        return path.as_posix()

adj = defaultdict(list)
imports_meta = {}
imported_by = defaultdict(list)
reachable = set()
queue = [entry.resolve()]
visited = set()

while queue:
    f = queue.pop()
    if f in visited:
        continue
    visited.add(f)
    if not f.exists() or f.suffix.lower() not in allowed_exts:
        continue
    reachable.add(f)
    text = f.read_text(encoding='utf-8', errors='ignore')
    matches = []
    for line in text.splitlines():
        stripped = line.strip()
        if not stripped.startswith(('import ', 'export ')):
            continue
        m = re.search(r'''['"]([^'"]+)['"]''', stripped)
        if not m:
            continue
        spec = m.group(1)
        if spec.endswith('.css'):
            continue
        resolved = resolve_import(f, spec)
        if resolved is None:
            matches.append((spec, None, 'unknown', stripped))
        else:
            is_type = 'import type' in stripped or 'import { type' in stripped or 'import type {' in stripped
            matches.append((spec, resolved, 'type' if is_type else 'runtime', stripped))
    imports_meta[f] = matches
    for spec, resolved, kind, line in matches:
        if resolved is None:
            continue
        if resolved not in file_set:
            continue
        adj[f].append(resolved)
        imported_by[resolved].append(f)
        if resolved not in visited:
            queue.append(resolved)

reachable = set(visited)

# circular imports
cycles = []
stack = []
stack_set = set()
seen = set()

def dfs(node):
    stack.append(node)
    stack_set.add(node)
    for nxt in adj.get(node, []):
        if nxt not in reachable:
            continue
        if nxt in stack_set:
            idx = stack.index(nxt)
            cycles.append(stack[idx:] + [nxt])
        elif nxt not in seen:
            dfs(nxt)
    stack.pop()
    stack_set.remove(node)
    seen.add(node)

for node in sorted(reachable):
    if node not in seen:
        dfs(node)

normalized_cycles = []
seen_cycles = set()
for cyc in cycles:
    if len(cyc) <= 1:
        norm = tuple(cyc)
    else:
        reps = [tuple(cyc[i:] + cyc[:i]) for i in range(len(cyc) - 1)]
        norm = min(reps)
    if norm not in seen_cycles:
        seen_cycles.add(norm)
        normalized_cycles.append([str(x) for x in norm])

# duplicates
interface_defs = defaultdict(list)
type_defs = defaultdict(list)
enum_defs = defaultdict(list)
constant_defs = defaultdict(list)
hook_defs = defaultdict(list)
system_defs = defaultdict(list)

for f in reachable:
    text = f.read_text(encoding='utf-8', errors='ignore')
    for m in re.finditer(r'\binterface\s+([A-Za-z0-9_]+)\b', text):
        interface_defs[m.group(1)].append(f)
    for m in re.finditer(r'\btype\s+([A-Za-z0-9_]+)\b', text):
        type_defs[m.group(1)].append(f)
    for m in re.finditer(r'\benum\s+([A-Za-z0-9_]+)\b', text):
        enum_defs[m.group(1)].append(f)
    for m in re.finditer(r'\b(?:const|let|var)\s+([A-Z0-9_]+)\b', text):
        constant_defs[m.group(1)].append(f)
    for m in re.finditer(r'\b(?:function|const)\s+(use[A-Za-z0-9_]+)\b', text):
        hook_defs[m.group(1)].append(f)
    for m in re.finditer(r'\b(?:class|function|const|let|var)\s+([A-Za-z0-9_]+(?:System|Game)[A-Za-z0-9_]*)\b', text):
        system_defs[m.group(1)].append(f)

# export info
export_pattern = re.compile(r'\bexport\s+(?:default\s+)?(?:async\s+)?(?:function|class|const|let|var|interface|type|enum)\s+([A-Za-z0-9_]+)')
export_named_pattern = re.compile(r'\bexport\s*\{([^}]+)\}')
export_default_pattern = re.compile(r'\bexport\s+default\b')

exports_by_file = {}
for f in reachable:
    text = f.read_text(encoding='utf-8', errors='ignore')
    exports = []
    for m in export_pattern.finditer(text):
        exports.append(('decl', m.group(1)))
    for m in export_named_pattern.finditer(text):
        inner = m.group(1)
        for part in inner.split(','):
            part = part.strip()
            if not part:
                continue
            if ' as ' in part:
                exports.append(('named', part.split(' as ')[1].strip()))
            else:
                exports.append(('named', part))
    if export_default_pattern.search(text):
        exports.append(('default', 'default'))
    exports_by_file[f] = exports

# unused exports
unused_exports = []
for f, exports in exports_by_file.items():
    for kind, name in exports:
        imported = False
        for src in reachable:
            if src == f:
                continue
            for spec, resolved, dep_kind, line in imports_meta.get(src, []):
                if resolved != f:
                    continue
                clause = line.split('from', 1)[0].replace('import ', '', 1).replace('type ', '', 1)
                if '{' in clause:
                    inner = clause.split('{', 1)[1].split('}', 1)[0]
                    for part in inner.split(','):
                        part = part.strip()
                        if ' as ' in part:
                            imported_name = part.split(' as ')[1].strip()
                        else:
                            imported_name = part.strip()
                        if imported_name == name or (name == 'default' and imported_name == 'default'):
                            imported = True
                            break
                elif clause.strip() and clause.strip() != '*':
                    if name == 'default':
                        imported = True
                if imported:
                    break
            if imported:
                break
        if not imported:
            unused_exports.append((f, kind, name))

# dead files/folders
all_imported_targets = set()
for src in reachable:
    for spec, resolved, kind, line in imports_meta.get(src, []):
        if resolved is not None and resolved in file_set:
            all_imported_targets.add(resolved)

dead_files = [f for f in all_code_files if f != entry.resolve() and f not in all_imported_targets]

all_dirs = defaultdict(list)
for f in all_code_files:
    all_dirs[f.parent].append(f)
dead_folders = []
for d, files in all_dirs.items():
    if any(f in reachable for f in files):
        continue
    rel_parts = d.relative_to(root).parts if d != root else ()
    if rel_parts and rel_parts[0] in {'.next', 'node_modules'}:
        continue
    dead_folders.append((d, [relpath(f) for f in files]))

broken_imports = []
for f in sorted(reachable):
    for spec, resolved, kind, line in imports_meta.get(f, []):
        if resolved is None:
            broken_imports.append((f, spec, line))

out_lines = []
out_lines.append('RUNTIME_DEPENDENCY_AUDIT')
out_lines.append('ENTRY=' + str(entry.resolve()))
out_lines.append('')
out_lines.append('1. COMPLETE DEPENDENCY TREE')
for f in sorted(reachable, key=lambda p: relpath(p)):
    out_lines.append('FILE=' + relpath(f))
    out_lines.append('ABSOLUTE_PATH=' + str(f))
    out_lines.append('IMPORTED_BY=' + ', '.join(relpath(x) for x in imported_by.get(f, [])) if imported_by.get(f) else 'NONE')
    imports_list = []
    for spec, resolved, kind, line in imports_meta.get(f, []):
        if resolved is None:
            imports_list.append(spec + ' [UNRESOLVED]')
        elif resolved in reachable:
            imports_list.append(relpath(resolved))
        else:
            imports_list.append(spec)
    out_lines.append('IMPORTS=' + ', '.join(imports_list) if imports_list else 'NONE')
    out_lines.append('RUNTIME_USED=YES')
    out_lines.append('')

out_lines.append('2. FILES THAT EXIST BUT ARE NEVER IMPORTED ANYWHERE')
for f in sorted(dead_files, key=lambda p: relpath(p)):
    out_lines.append(relpath(f))
out_lines.append('')

out_lines.append('3. EXPORTED SYMBOLS THAT ARE NEVER IMPORTED')
for f, kind, name in sorted(unused_exports, key=lambda x: (relpath(x[0]), x[1], x[2])):
    out_lines.append(relpath(f) + ' :: ' + kind + ' :: ' + name)
out_lines.append('')

out_lines.append('4. DUPLICATE INTERFACES')
for name, files in sorted(interface_defs.items()):
    if len(set(files)) > 1:
        out_lines.append(name)
        for f in sorted(set(files), key=lambda p: relpath(p)):
            out_lines.append('  ' + relpath(f))
out_lines.append('')

out_lines.append('5. DUPLICATE TYPES')
for name, files in sorted(type_defs.items()):
    if len(set(files)) > 1:
        out_lines.append(name)
        for f in sorted(set(files), key=lambda p: relpath(p)):
            out_lines.append('  ' + relpath(f))
out_lines.append('')

out_lines.append('6. DUPLICATE CONSTANTS')
for name, files in sorted(constant_defs.items()):
    if len(set(files)) > 1:
        out_lines.append(name)
        for f in sorted(set(files), key=lambda p: relpath(p)):
            out_lines.append('  ' + relpath(f))
out_lines.append('')

out_lines.append('7. DUPLICATE HOOKS')
for name, files in sorted(hook_defs.items()):
    if len(set(files)) > 1:
        out_lines.append(name)
        for f in sorted(set(files), key=lambda p: relpath(p)):
            out_lines.append('  ' + relpath(f))
out_lines.append('')

out_lines.append('8. DUPLICATE GAME SYSTEMS')
for name, files in sorted(system_defs.items()):
    if len(set(files)) > 1:
        out_lines.append(name)
        for f in sorted(set(files), key=lambda p: relpath(p)):
            out_lines.append('  ' + relpath(f))
out_lines.append('')

out_lines.append('9. CIRCULAR DEPENDENCY CHAINS')
for cyc in normalized_cycles:
    out_lines.append(' -> '.join(cyc))
out_lines.append('')

out_lines.append('10. FINAL RUNTIME GRAPH')
for f in sorted(reachable, key=lambda p: relpath(p)):
    targets = []
    for spec, resolved, kind, line in imports_meta.get(f, []):
        if resolved is not None and resolved in reachable and kind == 'runtime':
            targets.append(relpath(resolved))
    out_lines.append(relpath(f) + ' -> ' + ', '.join(targets) if targets else relpath(f) + ' -> NONE')
out_lines.append('')

out_lines.append('11. BROKEN IMPORTS / UNRESOLVED PATHS')
for f, spec, line in broken_imports:
    out_lines.append(relpath(f) + ' -> ' + spec + ' [UNRESOLVED]')
out_lines.append('')

out_lines.append('12. DEAD FOLDERS')
for d, files in sorted(dead_folders, key=lambda x: relpath(x[0])):
    out_lines.append(relpath(d) + ' :: ' + ', '.join(files))

out_path = root / 'runtime_audit_report.txt'
out_path.write_text('\n'.join(out_lines), encoding='utf-8')
print(out_path)
PY
python inspect_audit.py